File: backend/src/services/keycloak.service.ts
import axios, { AxiosInstance } from 'axios';
import config from '../config';
import logger from '../utils/logger';

/**
 * Keycloak Admin Service
 *
 * Handles all interactions with the Keycloak Admin API for:
 * - Realm management (multi-tenant provisioning)
 * - User management within realms
 * - Client and role configuration
 * - Identity provider setup
 */
class KeycloakService {
  private client: AxiosInstance;
  private accessToken: string | null = null;
  private tokenExpiry: number = 0;

  constructor() {
    this.client = axios.create({
      baseURL: config.keycloak.adminUrl,
      timeout: 10000,
    });
  }

  /**
   * Obtains an admin access token from Keycloak master realm.
   * Implements token caching to reduce authentication overhead.
   */
  private async getAdminToken(): Promise<string> {
    // Return cached token if still valid (with 30s buffer)
    if (this.accessToken && Date.now() < this.tokenExpiry - 30000) {
      return this.accessToken;
    }

    try {
      const response = await this.client.post(
        '/realms/master/protocol/openid-connect/token',
        new URLSearchParams({
          grant_type: 'password',
          client_id: 'admin-cli',
          username: config.keycloak.adminUser,
          password: config.keycloak.adminPassword,
        }),
        {
          headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
        }
      );

      this.accessToken = response.data.access_token;
      this.tokenExpiry = Date.now() + response.data.expires_in * 1000;

      logger.debug('Keycloak admin token obtained');
      return this.accessToken!;
    } catch (error) {
      logger.error('Failed to obtain Keycloak admin token', { error });
      throw new Error('Failed to authenticate with Keycloak');
    }
  }

  /**
   * Makes an authenticated request to the Keycloak Admin API.
   */
  private async adminRequest<T>(
    method: 'get' | 'post' | 'put' | 'delete',
    path: string,
    data?: any
  ): Promise<T> {
    const token = await this.getAdminToken();

    try {
      const response = await this.client.request<T>({
        method,
        url: `/admin/realms${path}`,
        data,
        headers: {
          Authorization: `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
      });
      return response.data;
    } catch (error: any) {
      const message = error.response?.data?.errorMessage || error.message;
      logger.error('Keycloak admin request failed', { path, error: message });
      throw new Error(`Keycloak API error: ${message}`);
    }
  }

  /**
   * Creates a new Keycloak realm for a tenant.
   * Each tenant gets their own isolated realm for complete IAM separation.
   */
  async createRealm(realmName: string, displayName: string): Promise<void> {
    logger.info('Creating Keycloak realm', { realmName, displayName });

    const realmConfig = {
      realm: realmName,
      displayName: displayName,
      enabled: true,
      registrationAllowed: false,
      resetPasswordAllowed: true,
      editUsernameAllowed: false,
      bruteForceProtected: true,
      permanentLockout: false,
      maxFailureWaitSeconds: 900,
      minimumQuickLoginWaitSeconds: 60,
      waitIncrementSeconds: 60,
      quickLoginCheckMilliSeconds: 1000,
      maxDeltaTimeSeconds: 43200,
      failureFactor: 5,
      // Session settings
      ssoSessionIdleTimeout: 1800,
      ssoSessionMaxLifespan: 36000,
      accessTokenLifespan: 300,
      accessTokenLifespanForImplicitFlow: 900,
      // Password policy for SOC2 compliance
      passwordPolicy: 'length(8) and upperCase(1) and lowerCase(1) and digits(1) and specialChars(1)',
      // Login settings
      loginWithEmailAllowed: true,
      duplicateEmailsAllowed: false,
      verifyEmail: true,
      // Token settings
      defaultSignatureAlgorithm: 'RS256',
    };

    const token = await this.getAdminToken();

    try {
      await this.client.post('/admin/realms', realmConfig, {
        headers: {
          Authorization: `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
      });
      logger.info('Keycloak realm created successfully', { realmName });
    } catch (error: any) {
      if (error.response?.status === 409) {
        logger.warn('Realm already exists', { realmName });
        return;
      }
      throw error;
    }
  }

  /**
   * Creates the TaskHub client application within a realm.
   * Configures OAuth2/OIDC settings for the frontend application.
   */
  async createClient(realmName: string): Promise<{ clientId: string; clientSecret: string }> {
    logger.info('Creating Keycloak client', { realmName });

    const clientConfig = {
      clientId: config.keycloak.clientId,
      name: 'TaskHub Application',
      description: 'TaskHub Multi-Tenant Project Management Application',
      enabled: true,
      clientAuthenticatorType: 'client-secret',
      protocol: 'openid-connect',
      // Public client for SPA (no client secret required for auth)
      publicClient: true,
      // Standard OAuth2 flow settings
      standardFlowEnabled: true,
      implicitFlowEnabled: false,
      directAccessGrantsEnabled: true,
      serviceAccountsEnabled: false,
      // Redirect URIs for the frontend
      redirectUris: [
        'http://localhost:3000/*',
        'http://localhost:3000/callback',
      ],
      webOrigins: [
        'http://localhost:3000',
        '+', // Allow all origins that match redirect URIs
      ],
      // Token settings
      attributes: {
        'access.token.lifespan': '300',
        'oauth2.device.authorization.grant.enabled': 'false',
        'oidc.ciba.grant.enabled': 'false',
      },
      // Default scopes
      defaultClientScopes: ['openid', 'profile', 'email', 'roles'],
    };

    await this.adminRequest('post', `/${realmName}/clients`, clientConfig);

    // Get the created client to retrieve its ID
    const clients = await this.adminRequest<any[]>(
      'get',
      `/${realmName}/clients?clientId=${config.keycloak.clientId}`
    );

    const client = clients[0];
    logger.info('Keycloak client created', { realmName, clientId: client.id });

    return {
      clientId: client.id,
      clientSecret: client.secret || '',
    };
  }

  /**
   * Creates realm-level roles for RBAC.
   * TaskHub uses three primary roles: admin, manager, member.
   */
  async createRealmRoles(realmName: string): Promise<void> {
    logger.info('Creating realm roles', { realmName });

    const roles = [
      {
        name: 'admin',
        description: 'Tenant administrator with full access',
        composite: false,
      },
      {
        name: 'manager',
        description: 'Project manager with elevated permissions',
        composite: false,
      },
      {
        name: 'member',
        description: 'Standard team member',
        composite: false,
      },
    ];

    for (const role of roles) {
      try {
        await this.adminRequest('post', `/${realmName}/roles`, role);
        logger.debug('Role created', { realmName, role: role.name });
      } catch (error: any) {
        if (error.message.includes('409') || error.message.includes('already exists')) {
          logger.debug('Role already exists', { realmName, role: role.name });
        } else {
          throw error;
        }
      }
    }
  }

  /**
   * Creates an initial admin user for a new tenant.
   */
  async createUser(
    realmName: string,
    email: string,
    firstName: string,
    lastName: string,
    password: string,
    roles: string[] = ['admin']
  ): Promise<string> {
    logger.info('Creating user in realm', { realmName, email });

    const userConfig = {
      username: email,
      email: email,
      firstName: firstName,
      lastName: lastName,
      enabled: true,
      emailVerified: true, // Skip verification for demo purposes
      credentials: [
        {
          type: 'password',
          value: password,
          temporary: false,
        },
      ],
    };

    await this.adminRequest('post', `/${realmName}/users`, userConfig);

    // Get the created user
    const users = await this.adminRequest<any[]>(
      'get',
      `/${realmName}/users?email=${encodeURIComponent(email)}`
    );

    const user = users[0];
    if (!user) {
      throw new Error('User creation failed - user not found after creation');
    }

    // Assign roles to the user
    for (const roleName of roles) {
      const role = await this.adminRequest<any>(
        'get',
        `/${realmName}/roles/${roleName}`
      );

      await this.adminRequest(
        'post',
        `/${realmName}/users/${user.id}/role-mappings/realm`,
        [role]
      );
    }

    logger.info('User created with roles', { realmName, userId: user.id, roles });
    return user.id;
  }

  /**
   * Gets all users in a realm with optional filtering.
   */
  async getUsers(
    realmName: string,
    params?: { search?: string; max?: number; first?: number }
  ): Promise<any[]> {
    const queryParams = new URLSearchParams();
    if (params?.search) queryParams.set('search', params.search);
    if (params?.max) queryParams.set('max', params.max.toString());
    if (params?.first) queryParams.set('first', params.first.toString());

    const query = queryParams.toString();
    return this.adminRequest<any[]>(
      'get',
      `/${realmName}/users${query ? `?${query}` : ''}`
    );
  }

  /**
   * Configures an identity provider (IdP) for social login or enterprise SSO.
   */
  async configureIdentityProvider(
    realmName: string,
    providerType: 'google' | 'github' | 'oidc',
    config: {
      alias: string;
      clientId: string;
      clientSecret: string;
      issuer?: string;
    }
  ): Promise<void> {
    logger.info('Configuring identity provider', { realmName, providerType });

    const providerConfigs: Record<string, any> = {
      google: {
        alias: config.alias || 'google',
        providerId: 'google',
        enabled: true,
        trustEmail: true,
        storeToken: false,
        addReadTokenRoleOnCreate: false,
        firstBrokerLoginFlowAlias: 'first broker login',
        config: {
          clientId: config.clientId,
          clientSecret: config.clientSecret,
          defaultScope: 'openid profile email',
        },
      },
      github: {
        alias: config.alias || 'github',
        providerId: 'github',
        enabled: true,
        trustEmail: true,
        storeToken: false,
        config: {
          clientId: config.clientId,
          clientSecret: config.clientSecret,
        },
      },
      oidc: {
        alias: config.alias,
        providerId: 'oidc',
        enabled: true,
        trustEmail: true,
        config: {
          clientId: config.clientId,
          clientSecret: config.clientSecret,
          tokenUrl: `${config.issuer}/protocol/openid-connect/token`,
          authorizationUrl: `${config.issuer}/protocol/openid-connect/auth`,
          userInfoUrl: `${config.issuer}/protocol/openid-connect/userinfo`,
          issuer: config.issuer,
        },
      },
    };

    await this.adminRequest(
      'post',
      `/${realmName}/identity-provider/instances`,
      providerConfigs[providerType]
    );
  }

  /**
   * Deletes a realm (tenant removal).
   */
  async deleteRealm(realmName: string): Promise<void> {
    logger.warn('Deleting Keycloak realm', { realmName });
    await this.adminRequest('delete', `/${realmName}`);
  }

  /**
   * Gets realm configuration.
   */
  async getRealm(realmName: string): Promise<any> {
    return this.adminRequest<any>('get', `/${realmName}`);
  }

  /**
   * Lists all realms (for admin purposes).
   */
  async listRealms(): Promise<any[]> {
    return this.adminRequest<any[]>('get', '');
  }
}

export const keycloakService = new KeycloakService();
export default keycloakService;

File: admin-service/app/services/keycloak_service.py
"""Keycloak administration service for tenant provisioning."""

import logging
from typing import Optional
import httpx

from app.config import settings

logger = logging.getLogger(__name__)


class KeycloakService:
    """Service for managing Keycloak realms and users."""

    def __init__(self):
        self.base_url = settings.keycloak_url
        self.admin_user = settings.keycloak_admin_user
        self.admin_password = settings.keycloak_admin_password
        self._token: Optional[str] = None

    async def _get_admin_token(self) -> str:
        """Get admin access token from Keycloak."""
        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{self.base_url}/realms/master/protocol/openid-connect/token",
                data={
                    "grant_type": "password",
                    "client_id": "admin-cli",
                    "username": self.admin_user,
                    "password": self.admin_password,
                },
            )
            response.raise_for_status()
            data = response.json()
            return data["access_token"]

    async def _get_headers(self) -> dict:
        """Get authorization headers."""
        token = await self._get_admin_token()
        return {
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json",
        }

    async def create_realm(
        self,
        realm_name: str,
        display_name: str,
    ) -> None:
        """Create a new Keycloak realm for a tenant."""
        logger.info(f"Creating Keycloak realm: {realm_name}")

        headers = await self._get_headers()

        realm_config = {
            "realm": realm_name,
            "enabled": True,
            "displayName": display_name,
            "registrationAllowed": False,
            "resetPasswordAllowed": True,
            "loginWithEmailAllowed": True,
            "duplicateEmailsAllowed": False,
            "sslRequired": "external",
            "roles": {
                "realm": [
                    {"name": "admin", "description": "Tenant administrator"},
                    {"name": "manager", "description": "Project manager"},
                    {"name": "member", "description": "Team member"},
                ]
            },
        }

        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{self.base_url}/admin/realms",
                headers=headers,
                json=realm_config,
            )
            response.raise_for_status()

        # Create the taskhub-app client
        await self._create_client(realm_name, headers)

        logger.info(f"Keycloak realm created: {realm_name}")

    async def _create_client(self, realm_name: str, headers: dict) -> None:
        """Create the taskhub-app client in the realm."""
        client_config = {
            "clientId": "taskhub-app",
            "enabled": True,
            "publicClient": True,
            "directAccessGrantsEnabled": True,
            "standardFlowEnabled": True,
            "implicitFlowEnabled": False,
            "redirectUris": ["http://localhost:3000/*"],
            "webOrigins": ["+", "http://localhost:3000"],
            "protocol": "openid-connect",
            "attributes": {
                "pkce.code.challenge.method": "S256",
            },
        }

        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{self.base_url}/admin/realms/{realm_name}/clients",
                headers=headers,
                json=client_config,
            )
            response.raise_for_status()

    async def create_user(
        self,
        realm_name: str,
        email: str,
        first_name: str,
        last_name: str,
        password: str,
        roles: list[str],
    ) -> str:
        """Create a user in the specified realm."""
        logger.info(f"Creating Keycloak user: {email} in realm {realm_name}")

        headers = await self._get_headers()

        user_config = {
            "username": email,
            "email": email,
            "firstName": first_name,
            "lastName": last_name,
            "enabled": True,
            "emailVerified": True,
            "credentials": [
                {
                    "type": "password",
                    "value": password,
                    "temporary": False,
                }
            ],
            "realmRoles": roles,
        }

        async with httpx.AsyncClient() as client:
            # Create user
            response = await client.post(
                f"{self.base_url}/admin/realms/{realm_name}/users",
                headers=headers,
                json=user_config,
            )
            response.raise_for_status()

            # Get user ID from location header
            location = response.headers.get("Location", "")
            user_id = location.split("/")[-1]

            # Assign realm roles
            await self._assign_realm_roles(realm_name, user_id, roles, headers)

            logger.info(f"Keycloak user created: {email} ({user_id})")
            return user_id

    async def _assign_realm_roles(
        self,
        realm_name: str,
        user_id: str,
        roles: list[str],
        headers: dict,
    ) -> None:
        """Assign realm roles to a user."""
        async with httpx.AsyncClient() as client:
            # Get available realm roles
            response = await client.get(
                f"{self.base_url}/admin/realms/{realm_name}/roles",
                headers=headers,
            )
            response.raise_for_status()
            available_roles = response.json()

            # Find matching roles
            roles_to_assign = [
                {"id": r["id"], "name": r["name"]}
                for r in available_roles
                if r["name"] in roles
            ]

            if roles_to_assign:
                # Assign roles to user
                response = await client.post(
                    f"{self.base_url}/admin/realms/{realm_name}/users/{user_id}/role-mappings/realm",
                    headers=headers,
                    json=roles_to_assign,
                )
                response.raise_for_status()

    async def delete_user(self, realm_name: str, user_id: str) -> None:
        """Delete a user from a realm."""
        headers = await self._get_headers()

        async with httpx.AsyncClient() as client:
            response = await client.delete(
                f"{self.base_url}/admin/realms/{realm_name}/users/{user_id}",
                headers=headers,
            )
            response.raise_for_status()

        logger.info(f"Keycloak user deleted: {user_id}")

    async def get_user_by_email(
        self, realm_name: str, email: str
    ) -> Optional[dict]:
        """Find a user by email in a realm."""
        headers = await self._get_headers()

        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"{self.base_url}/admin/realms/{realm_name}/users",
                headers=headers,
                params={"email": email, "exact": "true"},
            )
            response.raise_for_status()
            users = response.json()
            return users[0] if users else None

    async def delete_realm(self, realm_name: str) -> None:
        """Delete a Keycloak realm."""
        logger.info(f"Deleting Keycloak realm: {realm_name}")

        headers = await self._get_headers()

        async with httpx.AsyncClient() as client:
            response = await client.delete(
                f"{self.base_url}/admin/realms/{realm_name}",
                headers=headers,
            )
            # Ignore 404 if realm doesn't exist
            if response.status_code != 404:
                response.raise_for_status()

        logger.info(f"Keycloak realm deleted: {realm_name}")

    async def realm_exists(self, realm_name: str) -> bool:
        """Check if a realm exists."""
        headers = await self._get_headers()

        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"{self.base_url}/admin/realms/{realm_name}",
                headers=headers,
            )
            return response.status_code == 200


keycloak_service = KeycloakService()
